lib/foreman/gettext/support.rb

```diff
@@ -42,19 +42,31 @@ def self.add_text_domain(locale_domain, locale_dir)
           :report_warning => false
       end
 
-      def self.register_human_localenames
-        original_locale = FastGettext.locale
-        FastGettext.class.class_eval { attr_accessor :human_available_locales }
-        FastGettext.human_available_locales = []
-        FastGettext.default_available_locales.sort.each do |locale|
-          FastGettext.locale = locale
-          # TRANSLATORS: Provide locale name in native language (e.g. English, Deutsch or Portugues)
-          human_locale = _("locale_name")
-          human_locale = locale if human_locale == "locale_name"
-          FastGettext.human_available_locales << [human_locale, locale]
+      # Loading all available locales to get the language translation is slow
+      # (2 seconds for 20 languages), so let's lazy load that.
+      FastGettext.class.class_eval do
+        @@load_mutex = Mutex.new
+        @@human_available_locales = nil
+
+        def human_available_locales
+          return @@human_available_locales if @@human_available_locales
+          Rails.logger.debug "Loading and caching all available locales"
+          original_locale = FastGettext.locale
+          @@load_mutex.synchronize do
+            @@human_available_locales = []
+            FastGettext.default_available_locales.sort.each do |locale|
```

Comment:
Let's say we have threads A and B.
1) A calls human_available_locales, locks the mutex, sets the class variable to empty array and starts loading the locales
2) B calls human_available_locales, class variable is already set, B thinks only a subset of all actually available locales is available, depending on how far thread A got.
3) A finishes the loading and unlocks the mutex

Could we make the loading atomic so `@@human_available_locales` is `nil` up until it is fully populated? Then we would run into a different issue where the locales might get re-populated several times, depending on the timing, but I'd say that is a lesser issue than this.

Reply:
Nice catch, I was trying to avoid putting whole method into block because then I'd need to crate a new block for ensure. Lazy me. Everything needs to be synchronized indeed, well, Ruby don't have concurrent features built-in the language. It's 2021. But it does have tons of syntactic sugar... don't get me started. ;-)

For the record, `FastGettext.locale` is a thread-only global, so that can be outside. I can still keep ensure without begin, hurrah! :-)

Reply:
Btw next time just drop "unsafe threads" comment, you don't need to explain in that detail, it's obvious.